pandas/core/tools/datetimes.py

```diff
@@ -14,12 +14,18 @@
 from pandas.core.dtypes.common import (
     ensure_object, is_datetime64_dtype, is_datetime64_ns_dtype,
     is_datetime64tz_dtype, is_float, is_integer, is_integer_dtype,
-    is_list_like, is_numeric_dtype, is_object_dtype, is_scalar)
-from pandas.core.dtypes.generic import ABCDataFrame, ABCIndexClass, ABCSeries
+    is_list_like, is_numeric_dtype, is_scalar)
+from pandas.core.dtypes.generic import (
+    ABCDataFrame, ABCDatetimeIndex, ABCIndex, ABCIndexClass, ABCSeries)
 from pandas.core.dtypes.missing import notna
 
+from pandas._typing import ArrayLike, Union
 from pandas.core import algorithms
 
+# annotations
+DatetimeScalarOrArrayConvertible = Union[int, float, str, list, tuple,
```

Comment:
@gwrome has been working on a `DatetimeLike` Type in #26071 - do you see overlap here? Perhaps we should have a `Scalar` type in that module as well and do a union of `Scalar`, `DatetimeLike` and `ArrayLike` here 

Not a blocker just worth consideration

Reply:
I agree with a `Scalar` type so `Scalar` and `DatetimeScalar` unions were added by me.

As for a `DatetimeLike` type, it can contain a very large number of types, which makes it not applicable to our functions (and most likely almost anywhere). That is, it seems to me that we need to create a hierarchy that will contain, as base unions, `Union[int, float, str]`, `Union[Period, Timestamp, Timedelta]` and so on. And the top of this hierarchy will be a `DatetimeLike` union.

It is also not yet clear to me to which union to add a `list` and `tuple`.